components/support/webextensions/src/test/java/mozilla/components/support/webextensions/WebExtensionSupportTest.kt

```diff
@@ -404,6 +404,46 @@ class WebExtensionSupportTest {
         assertEquals(popupSessionId, (actionCaptor.value as TabListAction.RemoveTabAction).tabId)
     }
 
+    @Test
+    fun `allows overriding action popup behaviour`() {
+        val engine: Engine = mock()
+        val ext: WebExtension = mock()
+        val engineSession: EngineSession = mock()
+        val browserAction: Action = mock()
+        whenever(ext.id).thenReturn("test")
+        val store = spy(BrowserStore(BrowserState(
+            extensions = mapOf(ext.id to WebExtensionState(ext.id))
+        )))
+
+        val delegateCaptor = argumentCaptor<WebExtensionDelegate>()
+
+        var webExtensionId = ""
+        WebExtensionSupport.initialize(
+            engine,
+            store,
+            onActionPopupToggleOverride = {
+                webExtensionId = it.id
+            }
+        )
+        verify(engine).registerWebExtensionDelegate(delegateCaptor.capture())
+
+        // Toggling should allow state to have popup EngineSession instance
+        delegateCaptor.value.onToggleActionPopup(ext, engineSession, browserAction)
+        val actionCaptor = argumentCaptor<mozilla.components.browser.state.action.BrowserAction>()
+        verify(store, times(1)).dispatch(actionCaptor.capture())
+        val value = actionCaptor.value
+        assertNotNull((value as WebExtensionAction.UpdatePopupSessionAction).popupSession)
+        assertNotNull(store.state.extensions[webExtensionId]?.popupSession)
```

Comment:
You also need a `store.waitUntilIdle()` here otherwise this test will fail intermittenlty...